alg/viewshed.cpp

```diff
@@ -426,499 +480,389 @@ bool Viewshed::run(GDALRasterBandH hBand, GDALProgressFunc pfnProgress,
-            adfGeoTransform.data(), 1, 0, padfFirstLineVal[nX + 1], dfDistance2,
-            oOpts.curveCoeff, dfSphereDiameter));
-        pabyResult[nX + 1] = oOpts.visibleVal;
-        if (oOpts.outputMode != OutputMode::Normal)
-            dfHeightResult[nX + 1] = dfGroundLevel;
+        dfResult += (dfZ - dfCellVal);
+        dfResult = std::max(0.0, dfResult);
     }
+    else
+        dfResult = (dfCellVal + oOpts.targetHeight < dfZ) ? oOpts.invisibleVal
+                                                          : oOpts.visibleVal;
+    dfCellVal = std::max(dfCellVal, dfZ);
+}
+
+/// Process the first line (the one with the X coordinate the same as the observer).
+///
+/// @param nX  X location of the observer
+/// @param nY  Y location of the observer
+/// @param nLine  Line number being processed (should always be the same as nY)
+/// @param vLastLineVal  Vector in which to store the read line. Becomes the last line
+///    in further processing.
+/// @return True on success, false otherwise.
+bool Viewshed::processFirstLine(int nX, int nY, int nLine,
+                                std::vector<double> &vLastLineVal)
+{
+    int nYOffset = nLine - nY;  // Should be zero.
+    std::vector<double> vResult(oOutExtent.xSize());
+    std::vector<double> vThisLineVal(oOutExtent.xSize());
 
-    /* process left direction */
-    for (int iPixel = nX - 2; iPixel >= 0; iPixel--)
+    if (!readLine(nLine, vThisLineVal.data()))
+        return false;
+
+    // This bit is only relevant for the first line.
+    dfZObserver = oOpts.observer.z + vThisLineVal[nX];
+    dfHeightAdjFactor = CalcHeightAdjFactor(poDstDS.get(), oOpts.curveCoeff);
+    if (oOpts.outputMode == OutputMode::Normal)
     {
-        dfGroundLevel = 0;
-        if (oOpts.outputMode == OutputMode::DEM)
-            dfGroundLevel = padfFirstLineVal[iPixel];
+        vResult[nX] = oOpts.visibleVal;
+        if (nX - 1 >= 0)
+            vResult[nX - 1] = oOpts.visibleVal;
+        if (nX + 1 < oOutExtent.xSize())
+            vResult[nX + 1] = oOpts.visibleVal;
+    }
 
-        if (AdjustHeightInRange(adfGeoTransform.data(), nX - iPixel, 0,
-                                padfFirstLineVal[iPixel], dfDistance2,
-                                oOpts.curveCoeff, dfSphereDiameter))
-        {
-            double dfZ = CalcHeightLine(
-                nX - iPixel, padfFirstLineVal[iPixel + 1], dfZObserver);
+    // In DEM mode the base is the pre-adjustment value.
+    // In ground mode the base is zero.
+    if (oOpts.outputMode == OutputMode::DEM)
+        vResult = vThisLineVal;
 
-            if (oOpts.outputMode != OutputMode::Normal)
-                dfHeightResult[iPixel] = std::max(
-                    0.0, (dfZ - padfFirstLineVal[iPixel] + dfGroundLevel));
+    const auto [iLeft, iRight] =
+        adjustHeight(nYOffset, nX, vThisLineVal.data() + nX);
 
-            setVisibility(iPixel, dfZ, padfFirstLineVal, vResult);
-        }
-        else
+    std::thread t1(
```

Comment:
Is it worth the trouble threading the processing of a single line ... ?

Reply:
It was when I tested it. It's trivial to remove the threading and try it.